src/lib/redis/cluster.c

```diff
@@ -1472,6 +1480,31 @@ void *fr_redis_cluster_conn_create(TALLOC_CTX *ctx, void *instance, fr_time_delt
 		return NULL;
 	}
 
+#ifdef HAVE_REDIS_SSL
+	if (node->cluster->conf->ssl) {
+		redisSSLContextError ssl_error = REDIS_SSL_CTX_NONE;
+#ifdef HAVE_REDIS_SSL_OPTIONS
+		redisSSLOptions opt = { .verify_mode = REDIS_SSL_VERIFY_PEER };
+
+		ssl_context = redisCreateSSLContextWithOptions(&opt, &ssl_error);
+#else
+		ssl_context = redisCreateSSLContext(NULL, NULL, NULL, NULL, NULL, &ssl_error);
+#endif
+		if (!ssl_context || ssl_error != REDIS_SSL_CTX_NONE) {
+			ERROR("%s - [%i] Failed to create SSL context: %s", log_prefix, node->id, redisSSLContextGetError(ssl_error));
+			redisFree(handle);
+			return NULL;
+		}
+
+		if (redisInitiateSSLWithContext(handle, ssl_context) != REDIS_OK) {
```

Comment:
This could be `redisInitiateSSL(handle, SSL*)`, where the `SSL*` is taken from the FreeRADIUS APIs.

Reply:
Yeah, and then we get a bunch of configuration knobs and complex validation behaviour for free :)

Reply:
If you look through the Redis code, it's configuring the I/O layer for you, so all you have to do is pass in an SSL * with various configuration options set.

https://github.com/redis/hiredis/blob/master/ssl.c#L356

Reply:
So you suggest I just use SSL_new()?

Reply:
The thing we're looking for here is using the same configuration file layout as the other uses of TLS.  The reason why we don't use that function elsewhere is because other libraries (like libldap) don't expose the details of their TLS library, so we can't pass in an `SSL *`, and outbound RADSEC functionality hasn't been forward ported from v3 yet.

Reply:
As I said before, the redis code sets the FD, so that's not why the handshake is failing
```
static int redisSSLConnect(redisContext *c, SSL *ssl) {
    if (c->privctx) {
        __redisSetError(c, REDIS_ERR_OTHER, "redisContext was already associated");
        return REDIS_ERR;
    }

    redisSSL *rssl = hi_calloc(1, sizeof(redisSSL));
    if (rssl == NULL) {
        __redisSetError(c, REDIS_ERR_OOM, "Out of memory");
        return REDIS_ERR;
    }

    c->funcs = &redisContextSSLFuncs;
    rssl->ssl = ssl;

    SSL_set_mode(rssl->ssl, SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER);
    SSL_set_fd(rssl->ssl, c->fd);
    SSL_set_connect_state(rssl->ssl);
```

https://github.com/redis/hiredis/blob/master/ssl.c#L356

Reply:
If you look at the hiredis code, it's not doing anything special to prepare the `SSL *` https://github.com/redis/hiredis/blob/master/ssl.c#L415

Reply:
@arr2036 redis code not only prepares `SSL *` but also does `SSL_connect()` again (first time it is done in `fr_tls_session_alloc_client()` so I get `SSL_connect failed: Connection reset by peer` error when it is invoked for the second time

Reply:
Since nothing else uses `fr_tls_session_alloc_client()`, it's probably worth removing the call to `SSL_connect()` from that function.  The callers can then set up whatever IO handlers are necessary.

Please submit a separate PR for `fr_tls_session_alloc_client()`, which fixes the issue you found in it.  We can then have a separate PR for the redis SSL patches.

Reply:
@alandekok there is another issue: I can't use `SSL_CTX` created by `fr_tls_ctx_alloc()` because it sets msg, info and verify callbacks (which get called when redis tries to use `SSL` object created using that `SSL_CTX` and break because there are no any bound FR requests and such.

I do not see reasons to set these callbacks in `SSL_CTX` becasue the only users of `fr_tls_ctx_alloc()` are EAP modules which all use `eap_tls_session_init()` which in turn uses `fr_tls_session_alloc_server()` which sets own msg, info and verify callbacks.

So I would suggest to not set any callbacks for `SSL_CTX` in `fr_tls_ctx_alloc()`. What do you think?

Reply:
the `fr_tls_session_alloc_client()` function should probably not call `fr_tls_session_request_unbind(tls_session->ssl)` at the end of the function.  It's fine to have an internal (and otherwise unused) request associated with the client.

That way the callbacks will always have a `request` associated with them.  And it will be the same `request` for all of the callbacks.

Reply:
@alandekok do we really need these `SSL_CTX` callbacks for generic TLS clients? At least verify callback (as it is implemented now) breaks client TLS handshake for example.

I propose to keep the common code simpler:
- do not set up any callbacks when we create `SSL_CTX` - all the existing users of `fr_tls_ctx_alloc()` currently use `fr_tls_session_alloc_server()` which sets up necessary callbacks for them on its own (on `SSL` object)
- do not do `SSL_connect()` in `fr_tls_session_alloc_client()` like you suggested, or do it only when we pass in `fd` - this is the change that I have in my working tree

What do you think?

Reply:
Agree `SSL_connect()` should be removed from the alloc_client function.  If you want to ifdef the verify callbacks out and submit the PR that'd be fine with me.  Then we'll go back and figure out adding a fake request to the `SSL *` to run them later.

Reply:
@alandekok @arr2036 please review #5046 which contains src/lib/tls changes that are required